feat: add NEP-366 signDelegateAction support for meta-transactions

0xultravioleta opened the pull request:
## Background

This PR adds support for NEP-366 meta-transactions to enable the x402 payment protocol on NEAR. 

x402 is a decentralized payment standard that allows websites to monetize content using HTTP 402 Payment Required responses. For this to work seamlessly on NEAR, users need to be able to sign transactions without paying gas fees themselves - the relayer handles gas payment. This is exactly what NEP-366 DelegateActions provide.

By implementing `signDelegateAction`, MyNearWallet can now participate in the x402 ecosystem, allowing users to authorize gasless payments that relayers submit on their behalf.

## Live Demo

While this PR is pending review, we've deployed a version with this implementation:

**Wallet**: https://mynearwallet.ultravioletadao.xyz/

This deployment is currently being used by [402milly.xyz](https://402milly.xyz/) to demonstrate the implementation works in production.

## Summary
- Implements NEP-366 signDelegateAction with complete Borsh serialization
- Adds `/sign-delegate-action` route with approval UI
- Full Ledger hardware wallet support
- E2E tests (13 passing, 2 skipped that require testnet)
- **Tested in production** - Successfully completed x402 payment on mainnet

## Production Test

Tested end-to-end with [402milly.xyz](https://402milly.xyz/) - a pixel art marketplace using x402 payments.

| Component | Link |
|-----------|------|
| Facilitator service | [facilitator.ultravioletadao.xyz](https://facilitator.ultravioletadao.xyz) |
| Facilitator account (NEAR) | [uvd-facilitator.near](https://nearblocks.io/address/uvd-facilitator.near) |
| Storage deposit tx | [3oBFbvnUP277MPPDkzSLdcz4Tp9XjudebPqNaLcWmoM2](https://nearblocks.io/txns/3oBFbvnUP277MPPDkzSLdcz4Tp9XjudebPqNaLcWmoM2) |
| Payment tx | [FjaxvoxdfbXbstgPzjiRxdQmcjL2Q1vy6X5q3oGc9eqH](https://nearblocks.io/txns/FjaxvoxdfbXbstgPzjiRxdQmcjL2Q1vy6X5q3oGc9eqH) |

## Technical Implementation

### NEP-366 Signing
- **Prefix**: `(2^30 + 366) = 1073742190 = 0x4000016E` in little-endian
- **Signature**: ED25519 over `SHA256(prefix || Borsh(DelegateAction))`
- **Important**: `InMemorySigner.signMessage` already does SHA256 internally, so we pass raw data (prefix + delegateActionBytes), not pre-hashed

### Borsh Serialization
Complete implementation of NEP-366 binary format:
- `writeU8`, `writeU32`, `writeU64`, `writeU128` helpers using BigInt
- `serializePublicKey` - ED25519 key type byte + 32-byte key
- `serializeSignature` - ED25519 key type byte + 64-byte signature
- `serializeNonDelegateAction` - FunctionCall action serialization
- `serializeDelegateAction` - Full delegate action structure
- `serializeSignedDelegateAction` - DelegateAction + Signature

### Redirect Fix
Immediate redirect after signing to prevent RPC balance fetch errors that could crash the wallet before callback completes.

## Features
- **Gasless badge** - Shows users they won't pay gas fees
- **Ledger support** - Detects Ledger accounts, shows confirmation modal
- **Security** - Validates callback URLs against javascript: protocol
- **Cancel flow** - Redirects with userRejected error code

## Files Changed
- `packages/frontend/src/utils/wallet/delegateAction.ts` - Core signing utility with Borsh serialization
- `packages/frontend/src/routes/SignDelegateActionWrapper.js` - Route handler with immediate redirect
- `packages/frontend/src/components/sign-delegate-action/` - UI components
- `packages/e2e-tests/signDelegateAction/` - E2E tests

## Test plan
- [x] UI displays contract, actions, gasless badge
- [x] Cancel redirects with userRejected error
- [x] Invalid params show error messages
- [x] javascript: callback URLs rejected
- [x] Button states work during signing
- [x] Real x402 payment flow tested on mainnet
- [ ] Manual test with Ledger device

lostpebble commented:
Hi @0xultravioleta ,

Thanks for this PR. This is an important feature, but I get the impression a lot of this was "vibe coded" and it seems to be overlooking a few things. The code quality is not of a bad standard though- but it has some holes in it that are cause for concern, such as:

```ts
    // Create message to sign: prefix + sha256(serialized delegateAction)
    // Note: Using JSON serialization for simplicity. For full NEP-366 compliance,
    // this should use borsh serialization with the DelegateAction schema.
    const message = JSON.stringify({
        senderId: delegateAction.senderId,
        receiverId: delegateAction.receiverId,
        actions: actions.map((a) => ({
            methodName: a.methodName,
            args: a.args,
            gas: a.gas,
            deposit: a.deposit,
        })),
        nonce: delegateAction.nonce,
        maxBlockHeight: delegateAction.maxBlockHeight,
        publicKey: publicKey.toString(),
    });

    // Hash the message
    const messageHash = new Uint8Array(sha256.array(message));

    // Add NEP-366 prefix
    const prefixBuffer = new ArrayBuffer(4);
    new DataView(prefixBuffer).setUint32(0, DELEGATE_ACTION_PREFIX, true);
    const prefixArray = new Uint8Array(prefixBuffer);

    // Combine prefix and hash
    const dataToSign = new Uint8Array(prefixArray.length + messageHash.length);
    dataToSign.set(prefixArray);
    dataToSign.set(messageHash, prefixArray.length);
```

Did you verify that when this data is signed that it matches the signature of a `SignedDelegateAction` using the native libraries? I don't think we should be creating all of these things so manually- the NEAR JS libraries allow us to more easily create Delegate Action instances and sign them correctly.

We would like to support `DelegateAction` signing- but we need to also ensure we support it according to the spec defined by Wallet Selector as well. The interface defined by Wallet Selector is:

```ts
signDelegateAction(delegateAction: DelegateAction): Promise<[Uint8Array, SignedDelegate]>;
```

So, we need to be able to accept Borsh serialized Delegate Actions (base64 URL encoded preferably) and then decode them as real `DelegateAction` instances and use them in the wallet. The Delegate Action itself contains all the info required to sign it:

```ts
class DelegateAction {
    senderId: string;
    receiverId: string;
    actions: Array<Action>;
    nonce: bigint;
    maxBlockHeight: bigint;
    publicKey: PublicKey;
```

There is a case to be made for a Delegate Action signing mechanism like you've created here (where its more open ended and the user can pick their account etc.) and then we create the Delegate Action, like you've done. This does seem somewhat insecure though, as it allows anyone to create links for transactions without context and knowledge of the user's accounts (only providing a `receiverId` and `actions` array).

Basically, we need to support both methods- but the encoded Delegate Action (from dapp / wallet selector side) should be the priority. This creates a more intentional link, requiring user sign in / participation on the dapp side.

Initial support for Delegate Actions should start with a borsh encoded Delegate Action to be signed by the wallet, using a param such as `delegateActionBase64Url` and could include the `meta` param as you've implemented it here. This provides all the context required (selected account ID, receiver ID etc.) and aligns with the current expected interface from NEAR and Wallet Selector and provides more security to users.

---

In any case, thanks for the PR- even though I do think it was mostly AI code- there are some good things within it. But we need to ensure we stick to the spec as much as possible, and also make use of the provided NEAR libraries instead of manually creating things ourselves (something AI does a lot).

0xultravioleta commented:
Thanks for the detailed feedback @lostpebble. You're right that the initial implementation had issues. I'd love to work together on getting this right - x402 is gaining traction and MyNearWallet could be the first web wallet on NEAR to fully support meta-transactions.

### What's Been Fixed

**1. Borsh Serialization (Complete Rewrite)**

The code you quoted was from an earlier iteration. The current implementation uses proper Borsh binary serialization:

```typescript
// NEP-366 prefix: (2^30 + 366) = 1073742190
const DELEGATE_ACTION_PREFIX = 1073742190;

// Complete Borsh serialization helpers
function writeU8(value: number): Uint8Array { ... }
function writeU32(value: number): Uint8Array { ... }  // little-endian
function writeU64(value: number | string | bigint): Uint8Array { ... }
function writeU128(value: string | number | bigint): Uint8Array { ... }
function writeString(str: string): Uint8Array { ... }  // 4-byte length prefix + UTF-8
function writeVecU8(data: Uint8Array): Uint8Array { ... }

function serializeDelegateAction(...): Uint8Array {
    // Proper Borsh struct serialization:
    // sender_id: String, receiver_id: String, actions: Vec<NonDelegateAction>,
    // nonce: u64, max_block_height: u64, public_key: PublicKey
}

function serializeSignedDelegateAction(delegateActionBytes, signature): Uint8Array {
    // DelegateAction bytes + Signature (1 byte key type + 64 bytes)
}
```

**On using @near-js/transactions**: I looked into using the library directly but couldn't find exported functions for DelegateAction/SignedDelegateAction Borsh serialization. The library has the schemas internally but they don't seem to be exposed for external use. Happy to refactor if you can point me to the right exports - would definitely prefer using the official library over custom serialization.

**2. Wallet Selector Spec Compliance (NEW)**

Added support for the `delegateActionBase64` parameter per the Wallet Selector specification:

```typescript
// New functions for Wallet Selector flow
export function decodeDelegateActionForDisplay(delegateActionBase64: string): DecodedDelegateActionInfo
export async function signPrebuiltDelegateAction({ accountId, delegateActionBase64 }): Promise<SignedDelegateActionResult>
```

This enables two flows:
- **Wallet Selector flow**: dApp provides pre-built `delegateActionBase64`, wallet just signs
- **Open-ended flow**: dApp provides `receiverId` + `actions`, wallet builds and signs

**3. Validation Improvements**

Added runtime validation for gas/deposit values and access key nonce responses.

### Production Testing

Successfully tested with x402 payments on mainnet:
- **Live deployment**: https://mynearwallet.ultravioletadao.xyz
- **Test transaction**: [661xcy9s...](https://nearblocks.io/txns/661xcy9s2AuPQNUJ7vgVEU6jsTDRe2cz3QUxtEWj2HE2)

### On AI-Assisted Development

Yes, this was developed with AI assistance. There's nothing wrong with that - it's a legitimate way to develop software. What matters is that the implementation is correct and works in production.

Let me know if you have any other concerns or suggestions!
